packages/find-root/src/index.ts

```diff
@@ -31,50 +33,60 @@ async function hasWorkspacesConfiguredViaPkgJson(
   }
 }
 
-async function hasWorkspacesConfiguredViaLerna(directory: string) {
-  try {
-    let lernaJson = await fs.readJson(path.join(directory, "lerna.json"));
-    if (lernaJson.useWorkspaces !== true) {
-      return directory;
-    }
-  } catch (err) {
-    if (err.code !== "ENOENT") {
-      throw err;
-    }
-  }
-}
+export async function findRoot(cwd: string): Promise<MonorepoRoot> {
+  let monorepoRoot: MonorepoRoot | undefined;
+
+  await findUp(
+    async directory => {
+      return Promise.all(
+        defaultOrder.map(toolType => supportedTools[toolType]).map(async (tool: Tool): Promise<MonorepoRoot | undefined> => {
+          if (await tool.isMonorepoRoot(directory)) {
+            return {
+              tool: tool,
+              dir: directory
+            };
+          }
+        })
+      ).then(x => x.find(value => value)).then(result => {
+        if (result && !monorepoRoot) {
+          monorepoRoot = result;
+        }
 
-async function hasWorkspacesConfiguredViaPnpm(directory: string) {
-  // @ts-ignore
-  let pnpmWorkspacesFileExists = await fs.exists(
-    path.join(directory, "pnpm-workspace.yaml")
+        // the findUp tool expects us to return the current directory or undefined
+        return result ? result.dir : undefined;
+      });
+    },
+    { cwd, type: "directory" }
   );
-  if (pnpmWorkspacesFileExists) {
-    return directory;
-  }
-}
 
-export async function findRoot(cwd: string): Promise<string> {
+  if (monorepoRoot)
+    return monorepoRoot;
+
+  // No monorepo root was found for any supported tool, so instead we return... something??
+  // I'm not sure what the case for this "default to current root pkg if no monorepo found"
+  // behavior is, it feels like we should return undefined and the calling tool should know
+  // that there is no monorepo.
+
   let firstPkgJsonDirRef: { current: string | undefined } = {
```

Comment:
I know that you just touched this code but it's written really weirdly. Isn't this whole thing (from this moment to the end of the function) equivalent to smth like?
```ts
findUp(async directory => {
  const exists = await fs.exists(path.join(directory, 'package.json'))
  return exists ? directory : undefined
})
```

Reply:
Yes, it is written really strangely. In the original code before my edits, it was written that way to avoid a second findUp traversal.

I'll fix this up to match your suggested logic, because I think it is _inevitable_ that we need a second findUp if there's no monorepo tool found: especially later, when we allow users to pass in a filtered list of Tools, it just doesn't make sense to be looking for package.json files during the first traversal when the user's chosen tool might not even be "package.json"-based.